pallets/auctions/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+                Auctions::<T>::get(nft_id).ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+            // TODO : Refund any reserved bids
+
+            // List nft as not for sale
+            T::NFTHandler::set_listed_for_sale(nft_id, false)?;
+
+            // Remove auction from storage
+            Auctions::<T>::remove(nft_id);
+
+            // Emit auction canceled event
+            Self::deposit_event(Event::AuctionCancelled {
+                nft_id,
+                marketplace_id: current_auction.marketplace_id,
+                creator: who,
+            });
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        pub fn add_bid(
+            origin: OriginFor<T>,
+            nft_id: NFTId,
+            amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let current_block = frame_system::Pallet::<T>::block_number();
+
+            // fetch the data of given nftId
+            let nft_data = T::NFTHandler::get_nft(nft_id).ok_or(Error::<T>::NFTIdInvalid)?;
+
+            // fetch data of auction that lists NFT
+            let current_auction =
+                Auctions::<T>::get(nft_id).ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+            // ensure the nft is in listed for sale state
+            ensure!(
+                nft_data.listed_for_sale == false,
+                Error::<T>::NFTNotListedForSale
+            );
+
+            // ensure the caller is not the owner of NFT
+            ensure!(
+                nft_data.owner != who.clone(),
+                Error::<T>::OwnerCannotCreateBid
+            );
+
+            // ensure the auction period has commenced
+            ensure!(
+                current_auction.start_block < current_block,
+                Error::<T>::AuctionNotStarted
+            );
+
+            // ensure the auction period has not ended
+            ensure!(
+                current_auction.end_block > current_block,
+                Error::<T>::AuctionEnded
+            );
+
+            // ensure the bid is larger than the current highest bid
+            if let Some(current_highest_bid) = current_auction.bidders.get_current_highest_bid() {
+                ensure!(amount > current_highest_bid.1, Error::<T>::InvalidBidAmount);
+            }
+
+            // transfer funds from caller
+            T::CurrencyCaps::transfer(&who, &Self::account_id(), amount, KeepAlive)?;
+
+            // add bid to storage
+            Auctions::<T>::try_mutate(nft_id, |maybe_auction| -> DispatchResult {
+                // should not panic when unwrap since already checked above
+                let auction = maybe_auction.as_mut().unwrap();
+
+                // replace top bidder with caller
+                // if bidder has been removed, refund removed user
+                if let Some(bid) = auction.bidders.insert_new_bid(who.clone(), amount) {
```

Comment:
This should go after adding the auction grace period to the auction. 
Reason: First we do all things that can/might fail and then we do our magic :)